pkg/schedule/affinity/txn.go

```diff
@@ -0,0 +1,805 @@
+		if group.affinityVer != affinityVer {
+			return group, nil
+		}
+		// Group must not change voterStoreIDs while it is not in the expired status.
+		// RegularSchedulingEnabled == IsExpired
+		// The VoterStoreIDs are already sorted, so they can be compared directly
+		if !group.RegularSchedulingEnabled && !slices.Equal(voterStoreIDs, group.VoterStoreIDs) {
+			return group, nil
+		}
+	}
+
+	// Step 2: Save the Group in storage.
+	group.LeaderStoreID = leaderStoreID
+	group.VoterStoreIDs = slices.Clone(voterStoreIDs)
+	if err := m.storage.RunInTxn(m.ctx, func(txn kv.Txn) error {
+		return m.storage.SaveAffinityGroup(txn, groupID, &group.Group)
+	}); err != nil {
+		return nil, err
+	}
+
+	// Step 3: Save the information in memory.
+	return m.updateGroupPeers(groupID, leaderStoreID, voterStoreIDs)
+}
+
+// UpdateAffinityGroupKeyRanges batch modifies key ranges for multiple affinity groups.
+// Note: Validates add and remove operations separately, then applies both atomically in a single transaction.
+// Currently migration (removing a range from group A and adding the same range to group B in one request)
+// is not supported because add validation checks overlaps against existing ranges before removals are applied.
+func (m *Manager) UpdateAffinityGroupKeyRanges(addOps, removeOps []GroupKeyRanges) error {
+	toAdd := make(map[string]GroupKeyRanges, len(addOps))
+	toRemove := make(map[string]GroupKeyRanges, len(removeOps))
+	newAddedLabelRules := make(map[string]*labeler.LabelRule, len(addOps))
+	newRemovedLabelRules := make(map[string]*labeler.LabelRule, len(removeOps))
+
+	plan := m.regionLabeler.NewPlan()
+
+	// Step 0: Validate that a Group is either fully added or fully removed.
+	for _, op := range addOps {
+		if len(op.KeyRanges) == 0 {
+			continue
+		}
+		if _, exists := toAdd[op.GroupID]; exists {
+			return errs.ErrAffinityGroupExist.GenWithStackByArgs(op.GroupID)
+		}
+		toAdd[op.GroupID] = GroupKeyRanges{}
+	}
+	for _, op := range removeOps {
+		if len(op.KeyRanges) == 0 {
+			continue
+		}
+		if _, exists := toRemove[op.GroupID]; exists {
+			return errs.ErrAffinityGroupExist.GenWithStackByArgs(op.GroupID)
+		}
+		if _, exists := toAdd[op.GroupID]; exists {
+			return errs.ErrAffinityGroupExist.GenWithStackByArgs(op.GroupID)
+		}
+		toRemove[op.GroupID] = GroupKeyRanges{}
+	}
+
+	m.metaMutex.Lock()
+	defer m.metaMutex.Unlock()
+
+	// Step 1: Validate the added KeyRanges.
+	var newRangesToValidate []GroupKeyRanges
+	for _, op := range addOps {
+		currentRanges, err := m.getCurrentRanges(op.GroupID)
+		if err != nil {
+			return err
+		}
+
+		// Merge current ranges with new ranges to add
+		toAdd[op.GroupID] = GroupKeyRanges{
+			GroupID:   op.GroupID,
+			KeyRanges: slices.Concat(currentRanges.KeyRanges, op.KeyRanges),
+		}
+
+		// Collect new ranges for overlap validation
+		newRangesToValidate = append(newRangesToValidate, GroupKeyRanges{
```

Comment:
Why we need a new structure? It is same as `addOps`

Reply:
I will fix it.